engine/src/main/scala/cromwell/engine/workflow/WorkflowStoreActor.scala

```diff
@@ -0,0 +1,145 @@
+    workflowStore ++= submittedWorkflows
+    submittedWorkflows map { _._1 }
+  }
+
+  /**
+    * Retrieves up to n workflows which have not already been pulled into the engine and sets their pickedUp
+    * flag to true
+    */
+  def fetchRunnableWorkflows(n: Int): immutable.List[WorkflowToStart] = {
+    val startableWorkflows = workflowStore filter { case (k, v) => v.state.isStartable } take n
+    workflowStore ++= startableWorkflows map { case (k, v) => k -> SubmittedWorkflow(v.sources, Running) }
+    // Because we explicitly filtered out for startable states this is safe
+    startableWorkflows map { case (k, v) => v.toWorkflowToStart(k) } toList
+  }
+
+  // FIXME: Do we care if it doesn't exist? If so to what extent
+  def remove(id: WorkflowId): Unit = workflowStore -= id
+
+  protected def dump: immutable.Map[WorkflowId, SubmittedWorkflow] = workflowStore.toMap
+}
+
+class WorkflowStoreActor extends WorkflowStore with Actor with ServiceRegistryClient {
+  /*
+    WARNING: WorkflowStore is NOT thread safe. Unless that statement is no longer true do NOT use threads
+    outside of the the single threaded Actor event loop
+   */
+
+  private val logger = Logging(context.system, this)
+
+  override def receive = {
+    case SubmitWorkflow(source) =>
+      val id = add(source)
+      sendIdToMetadataService(id)
+      sender ! WorkflowAdded(id)
+    case SubmitWorkflows(sources) =>
+      val ids = add(sources)
+      ids foreach sendIdToMetadataService
+      sender ! WorkflowsAdded(add(sources))
+    case FetchRunnableWorkflows(n) => sender ! NewWorkflows(fetchRunnableWorkflows(n))
+    case RemoveWorkflow(id) => remove(id)
+  }
+
+  /**
+    * Takes the workflow id and sends it over to the metadata service w// default empty values for inputs/outputs
+    */
+  private def sendIdToMetadataService(id: WorkflowId): Unit = {
+    val submissionEvents = List(
+      MetadataEvent(MetadataKey(id, None, WorkflowMetadataKeys.SubmissionTime), MetadataValue(OffsetDateTime.now.toString)),
+      MetadataEvent.empty(MetadataKey(id, None, WorkflowMetadataKeys.Inputs)),
+      MetadataEvent.empty(MetadataKey(id, None, WorkflowMetadataKeys.Outputs))
+    )
+
+    serviceRegistryActor ! PutMetadataAction(submissionEvents)
+  }
+}
+
+object WorkflowStoreActor {
+  sealed trait WorkflowStoreActorCommand
+  case class SubmitWorkflow(source: WorkflowSourceFiles) extends WorkflowStoreActorCommand
+  case class SubmitWorkflows(sources: Seq[WorkflowSourceFiles]) extends WorkflowStoreActorCommand
+  case class FetchRunnableWorkflows(n: Int) extends WorkflowStoreActorCommand
+  case class RemoveWorkflow(id: WorkflowId) extends WorkflowStoreActorCommand
+
+  sealed trait WorkflowStoreActorResponse
+  case class WorkflowAdded(workflowId: WorkflowId) extends WorkflowStoreActorResponse
+  case class WorkflowsAdded(workflowIds: List[WorkflowId]) extends WorkflowStoreActorResponse
```

Comment:
is this `List` mutable, or are all the other `immutable.List` qualifications unnecessary?